edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/ServiceIdentityTree.cs

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.Azure.Devices.Edge.Hub.Core.Identity.Service;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Azure.Devices.Edge.Util.Concurrency;
+
+    /// <summary>
+    /// This is a tree implementation that uses the ParentScopes property
+    /// to reconstruct the hierarchy of modules/devices under the current
+    /// Edge device in a nested Edge environment.
+    /// </summary>
+    public class ServiceIdentityTree :
+        IServiceIdentityTree,
+        IAuthenticationChainProvider
+    {
+        readonly string rootDeviceId;
+        AtomicReference<ImmutableDictionary<string, ServiceIdentityTreeNode>> nodes;
+
+        public ServiceIdentityTree(string rootDeviceId)
+        {
+            this.rootDeviceId = Preconditions.CheckNonWhiteSpace(rootDeviceId, nameof(rootDeviceId));
+            this.nodes = new AtomicReference<ImmutableDictionary<string, ServiceIdentityTreeNode>>(ImmutableDictionary<string, ServiceIdentityTreeNode>.Empty);
+        }
+
+        public void InsertOrUpdate(ServiceIdentity identity)
+        {
+            // There should always be a valid ServiceIdentity
+            Preconditions.CheckNotNull(identity, nameof(identity));
+
+            ImmutableDictionary<string, ServiceIdentityTreeNode> snapshot = this.nodes;
+            var writableSnapshot = new Dictionary<string, ServiceIdentityTreeNode>(snapshot);
+            if (snapshot.ContainsKey(identity.Id))
+            {
+                // Update case - this is just remove + re-insert
+                this.Remove(identity.Id);
+            }
+
+            // Insert case
+            if (identity.IsModule)
+            {
+                this.InsertModuleIdentity(writableSnapshot, identity);
+            }
+            else
+            {
+                this.InsertDeviceIdentity(writableSnapshot, identity);
+            }
+
+            // Write back to the atomic field
+            if (!this.nodes.CompareAndSet(snapshot, writableSnapshot.ToImmutableDictionary()))
+            {
+                // There's only ever one thread calling InsertOrUpdate, so this should never fail
+                throw new InvalidOperationException("Multi-threaded insert/update detected on ServiceIdentityTree");
```

Comment:
How will throwing work here? Do you expect the caller to re-try? WHy not add that logic here?